[SYCL] Remove need to mark free functions with SYCL_EXTERNAL attribute

rdeodhar opened the pull request:
This change removes the need to mark free functions with the SYCL_EXTERNAL attribute. 

Within clang some instances of add_ir_attribute_function contain non-constexpr expressions (for example, introduced by sycl/ext/oneapi/matrix/matrix-unified.hpp). This occurs when there are non-instantiated functions/methods used in defining attribute values. These expressions are converted into constexprs later in the compilation pipeline.

Calling getAttributeNameValuePairs unconditionally leads to compile-time asserts. To account for the non-constexpr attribute values, a check is first made to rule out instances of the problematic values.

maarquitos14 reviewed (approved):
LGTM.

elizabethandrews reviewed (commented):
Please add FE tests for FE changes.

rdeodhar commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -16626,7 +16626,7 @@ Decl *Sema::ActOnFinishFunctionBody(Decl *dcl, Stmt *Body,
 
   // Handle free functions.
-  if (LangOpts.SYCLIsDevice && FD->hasAttr<SYCLDeviceAttr>() && Body &&
+  if (FD && Body &&
```
Wouldn't that defeat the purpose of this PR? It is trying to do away with the need to place SYCL_EXTERNAL on a free function definition. If I take out that condition then free functions without SYCL_EXTERNAL won't be processed as such.

elizabethandrews replied:
`LangOpts.SYCLIsDevice` does not mean `SYCL_EXTERNAL`. It just   means SYCL device compilation

rdeodhar replied:
Yes, you're right. Otherwise free function bodies would be placed in host code too. I've put back the check for `LangOpts.SYCLIsDevice.`

rdeodhar commented:
> Please add FE tests for FE changes.

I've updated the int_header test. It no longer uses the `__attribute__((sycl_device))` markup.

elizabethandrews commented:
> > Please add FE tests for FE changes.
> 
> I've updated the int_header test. It no longer uses the `__attribute__((sycl_device))` markup.

I don't think this is enough. You did this to fix crashes you saw when using dependent expressions. You need to add tests which illustrate this.

rdeodhar commented:
> > > Please add FE tests for FE changes.
> > 
> > 
> > I've updated the int_header test. It no longer uses the `__attribute__((sycl_device))` markup.
> 
> I don't think this is enough. You did this to fix crashes you saw when using dependent expressions. You need to add tests which illustrate this.

About 700 existing tests fail if the dependent expression check is removed.
What kind of test are you looking for?

elizabethandrews commented:
> > > > Please add FE tests for FE changes.
> > > 
> > > 
> > > I've updated the int_header test. It no longer uses the `__attribute__((sycl_device))` markup.
> > 
> > 
> > I don't think this is enough. You did this to fix crashes you saw when using dependent expressions. You need to add tests which illustrate this.
> 
> About 700 existing tests fail if the dependent expression check is removed. What kind of test are you looking for?

Are these FE tests? If not, I think you need a FE test which illustrates what you are trying to achieve with the following check. I don't fully know what you were trying to fix here - is it cases when attribute arguments are templated? You should add a test which shows this from a FE perspective. A case which didn't work before you added this check but now does so. Removing the sycl_device attribute does not illustrate what exactly you're trying to make work here.  

```
    if (hasDependentExpr(IRAttr->args_begin(), IRAttr->args_size()))
      continue;
```

rdeodhar commented:
> > > > > Please add FE tests for FE changes.
> > > > 
> > > > 
> > > > I've updated the int_header test. It no longer uses the `__attribute__((sycl_device))` markup.
> > > 
> > > 
> > > I don't think this is enough. You did this to fix crashes you saw when using dependent expressions. You need to add tests which illustrate this.
> > 
> > 
> > About 700 existing tests fail if the dependent expression check is removed. What kind of test are you looking for?
> 
> Are these FE tests? If not, I think you need a FE test which illustrates what you are trying to achieve with the following check. I don't fully know what you were trying to fix here - is it cases when attribute arguments are templated? You should add a test which shows this from a FE perspective. A case which didn't work before you added this check but now does so. Removing the sycl_device attribute does not illustrate what exactly you're trying to make work here.
> 
> ```
>     if (hasDependentExpr(IRAttr->args_begin(), IRAttr->args_size()))
>       continue;
> ```

Currently, the attribute SYCL_EXTERNAL, along with the free function property, is necessary for the function to be recognized as a free function. This requirement is reflected in this line in SemaDecl.cpp:
`if (LangOpts.SYCLIsDevice && FD->hasAttr<SYCLDeviceAttr>() && ...`
Unless the function is marked as SYCL_EXTERNAL it is not checked for having the free function attributes.

We want to remove the need to mark free functions as SYCL_EXTERNAL. Now, if the subexpression `&& FD->hasAttr<SYCLDeviceAttr>()`  in the above check is removed, many tests fail. Here is a sampling:

Failed Tests (346):
  Clang :: CodeGenSYCL/add_ir_attributes_function.cpp
  Clang :: CodeGenSYCL/add_ir_attributes_function_filter.cpp
  Clang :: Driver/clang-offload-bundler-bc-archive-support-linux-old-model.cpp
  SYCL :: abi/layout_array.cpp
  SYCL :: abi/preview_lib_marker.cpp
  SYCL :: abi/sycl_classes_abi_neutral_test.cpp
  SYCL :: abi/vtable.cpp
  SYCL :: aliases-opencl.cpp
...

The reason for the failures is the presence of some non-constexpr expressions in add_ir_function_attribute, most often introduced by a joint matrix header file. That causes a crash when the function `getAttributeNameValuePairs `is called during the check for free function attributes in SemaSYCL.cpp. Therefore we need to call `getAttributeNameValuePairs` only if non-constexprs are absent from the add_ir_attribute_function attribute values. Hence the changes in SemaSYCL.cpp to first check for dependent expressions, which excludes non-constepr expressions.

So given this background, what type of test do you suggest I add?

elizabethandrews commented:
>>Failed Tests (346):
Clang :: CodeGenSYCL/add_ir_attributes_function.cpp
Clang :: CodeGenSYCL/add_ir_attributes_function_filter.cpp

Can you paste the failing case here? If this sufficiently covers your change, you may not need additional tests

AaronBallman commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1085,12 +1085,19 @@ static target getAccessTarget(QualType FieldTy,
       AccTy->getTemplateArgs()[3].getAsIntegral().getExtValue());
 }
 
-// FIXME: Free functions must have void return type, be declared at file scope,
```
You've lost the part of the comment about requiring a `void` return type.

rdeodhar replied:
Fixed.

AaronBallman commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1085,12 +1085,19 @@ static target getAccessTarget(QualType FieldTy,
+// FIXME: Free functions must be declared at file scope, outside any namespaces.
 static bool isFreeFunction(SemaSYCL &SemaSYCLRef, const FunctionDecl *FD) {
   for (auto *IRAttr : FD->specific_attrs<SYCLAddIRAttributesFunctionAttr>()) {
+    // Free function properties are all compiletime constants, so skip checking
+    // any attribute values that are not constants.
+    if (hasDependentExpr(IRAttr->args_begin(), IRAttr->args_size()))
```
I'm not certain I understand this... dependence has nothing to do with whether something is or isn't constant. The general idea is: Sema checks some non-dependent conditions so it can give early diagnostics before a template is instantiated, but all the dependent conditions eventually become non-dependent once the template is instantiated and that's when the conditions are checked.

So I think this comment needs to be updated, or I'm misunderstanding something about the changes.

rdeodhar replied:
While searching for other uses of `getAttributeNameValuePairs `I discovered that it is prefaced with the hasDependentExpr check in a couple of other places. That worked in this case too, so I adopted it.

AaronBallman replied:
Yeah, I'm not saying the code is wrong, I'm saying the comment associated with the code is wrong.

AaronBallman commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1085,12 +1085,19 @@ static target getAccessTarget(QualType FieldTy,
-// outside any namespaces, and with the SYCL_DEVICE attribute. If the
-// SYCL_DEVICE attribute is not specified this function is not entered since the
-// possibility of the function being a free function is ruled out already.
+static bool hasDependentExpr(Expr **Exprs, const size_t ExprsSize) {
```
This is simple enough I'd probably make it a lambda along these lines:
```
auto hasDependentExpr = [](auto Args) {
  return llvm::any_of(Args, [](const Expr *E) {
    return E->isValueDependent() || E->isTypeDependent();
  });
};
```
(you can call it with `IRAttr->args()` to use the range-based APIs, which is cleaner)

rdeodhar replied:
OK, I've made that change.

AaronBallman reviewed (commented):
Why was SYCL_EXTERNAL necessary in the first place? That may help with judging what kind of test coverage to add. Or was this marking always entirely unnecessary except by spec?

rdeodhar commented:
> Why was SYCL_EXTERNAL necessary in the first place? That may help with judging what kind of test coverage to add. Or was this marking always entirely unnecessary except by spec?

Free functions are also normal callable functions in device code. Therefore in the first implementation it seemed reasonable to require that they be marked as SYCL_EXTERNAL. It also reduced the number of functions that are tested for being free functions. However, it would be desirable that a free function markup should _imply_ that the function is a device function. Therefore this PR is now trying to remove the requirement for SYCL_EXTERNAL.

AaronBallman commented:
> > Why was SYCL_EXTERNAL necessary in the first place? That may help with judging what kind of test coverage to add. Or was this marking always entirely unnecessary except by spec?
> 
> Free functions are also normal callable functions in device code. Therefore in the first implementation it seemed reasonable to require that they be marked as SYCL_EXTERNAL. It also reduced the number of functions that are tested for being free functions. However, it would be desirable that a free function markup should _imply_ that the function is a device function. Therefore this PR is now trying to remove the requirement for SYCL_EXTERNAL.

Thanks! Then I think what we need significantly more testing around are free functions which do things that are not allowed in a device function but are now available to be called within a device function to ensure we emit appropriate diagnostics and don't have assertion failures/crashes due to unexpected state.

rdeodhar commented:
> > > Why was SYCL_EXTERNAL necessary in the first place? That may help with judging what kind of test coverage to add. Or was this marking always entirely unnecessary except by spec?
> > 
> > 
> > Free functions are also normal callable functions in device code. Therefore in the first implementation it seemed reasonable to require that they be marked as SYCL_EXTERNAL. It also reduced the number of functions that are tested for being free functions. However, it would be desirable that a free function markup should _imply_ that the function is a device function. Therefore this PR is now trying to remove the requirement for SYCL_EXTERNAL.
> 
> Thanks! Then I think what we need significantly more testing around are free functions which do things that are not allowed in a device function but are now available to be called within a device function to ensure we emit appropriate diagnostics and don't have assertion failures/crashes due to unexpected state.

We already check functions marked with SYCL_EXTERNAL for doing device-compatible things. So we need to make the same checks for free functions as we do for SYCL_EXTERNAL functions.

rdeodhar commented:
> CodeGenSYCL/add_ir_attributes_function.cpp

One example of non-constexpr expressions in add_ir_attribute_function is clang/test/CodeGenSYCL/add_ir_attributes_function_filter.cpp:

```
template <typename... Properties>
#ifdef __SYCL_DEVICE_ONLY__
[[__sycl_detail__::add_ir_attributes_function(
    {"Prop4", "Prop3"},
    Properties::name..., Properties::value...)]]
#endif
void
free_func() {
}

template <typename... Properties>
class KernelFunctor {
public:
#ifdef __SYCL_DEVICE_ONLY__
  [[__sycl_detail__::add_ir_attributes_function(
      {"Prop3", "Prop5"},
      Properties::name..., Properties::value...)]]
#endif
  void
  operator()() const {
    free_func<Properties...>();
  }
};

int main() {
  sycl::queue q;
  auto f = [=]() {};
  q.submit([&](sycl::handler &h) {
    KernelFunctor<prop5, prop7, prop1, prop3, prop4> f{};
    h.single_task<class test_kernel>(f);
  });
}
```
Here, the properties prop3, prop4 and prop7, defined in mock_properties.hpp, are not constexpr strings.

AaronBallman commented:
> > > > Why was SYCL_EXTERNAL necessary in the first place? That may help with judging what kind of test coverage to add. Or was this marking always entirely unnecessary except by spec?
> > > 
> > > 
> > > Free functions are also normal callable functions in device code. Therefore in the first implementation it seemed reasonable to require that they be marked as SYCL_EXTERNAL. It also reduced the number of functions that are tested for being free functions. However, it would be desirable that a free function markup should _imply_ that the function is a device function. Therefore this PR is now trying to remove the requirement for SYCL_EXTERNAL.
> > 
> > 
> > Thanks! Then I think what we need significantly more testing around are free functions which do things that are not allowed in a device function but are now available to be called within a device function to ensure we emit appropriate diagnostics and don't have assertion failures/crashes due to unexpected state.
> 
> We already check functions marked with SYCL_EXTERNAL for doing device-compatible things. So we need to make the same checks for free functions as we do for SYCL_EXTERNAL functions. This is already being done through the "Checker" visitor.

You have no semantic test changes, only codegen and end to end tests. This means the code had to be sufficiently well-formed to get to the point of generating code, but your changes are all in Sema -- so it's not clear that the checker is actually doing anything. I would expect some test coverage showing a free function that isn't semantically valid and gets rejected.

rdeodhar commented:
> > > > > Why was SYCL_EXTERNAL necessary in the first place? That may help with judging what kind of test coverage to add. Or was this marking always entirely unnecessary except by spec?
> > > > 
> > > > 
> > > > Free functions are also normal callable functions in device code. Therefore in the first implementation it seemed reasonable to require that they be marked as SYCL_EXTERNAL. It also reduced the number of functions that are tested for being free functions. However, it would be desirable that a free function markup should _imply_ that the function is a device function. Therefore this PR is now trying to remove the requirement for SYCL_EXTERNAL.
> > > 
> > > 
> > > Thanks! Then I think what we need significantly more testing around are free functions which do things that are not allowed in a device function but are now available to be called within a device function to ensure we emit appropriate diagnostics and don't have assertion failures/crashes due to unexpected state.
> > 
> > 
> > We already check functions marked with SYCL_EXTERNAL for doing device-compatible things. So we need to make the same checks for free functions as we do for SYCL_EXTERNAL functions. This is already being done through the "Checker" visitor.
> 
> You have no semantic test changes, only codegen and end to end tests. This means the code had to be sufficiently well-formed to get to the point of generating code, but your changes are all in Sema -- so it's not clear that the checker is actually doing anything. I would expect some test coverage showing a free function that isn't semantically valid and gets rejected.

I've added a test free_function_errors.cpp which checks errors generated for invalid parameter types: struct, union, reference, array, SYCL special types.

rdeodhar commented:
I added the test in response to Ballman's comment: "Then I think what we need significantly more testing around are free functions which do things that are not allowed in a device function but are now available to be called within a device function to ensure we emit appropriate diagnostics and don't have assertion failures/crashes due to unexpected state."

That comment is unrelated to this PR but I added the test anyway to meet that request.

I still don't know how to meet your request to add a test. Please tell me what that test should do. Because this PR adds a check to make sure we don't trip over non-constexpr values in add_ir_attribute_function, there is no way to make it fail. 

This PR is not fixing a problem. If it were, then a good test would be one that has the problem, and the test would pass because the problem has been fixed.

This PR is making an **enhancement**. We no longer need to mark free functions with SYCL_EXTERNAL. And the test to prove that this enhancement works is the **removal of SYCL_EXTERNAL from some existing tests**.

elizabethandrews commented:
> This PR is not fixing a problem. If it were, then a good test would be one that has the problem, and the test would pass because the problem has been fixed. 
> 
> This PR is making an **enhancement**. We no longer need to mark free functions with SYCL_EXTERNAL. And the test to prove that this enhancement works is the **removal of SYCL_EXTERNAL from some existing tests**.

From the FE perspective, the PR is fixing a problem.  It is not just making an enhancement. You needed to mark free functions with SYCL_EXTERNAL earlier because of a bug/problem you have now fixed in this PR.  If you weren't fixing a bug, the only code you would have had in this PR would have been to remove the check for SYCL_EXTERNAL. 

You have added checks for dependent values. We don't fully understand what those checks have to do with non-constexpr values, because dependency and constexpr are not the same thing. Without a test we do not understand what FE bug you fixed here. 

> I still don't know how to meet your request to add a test. Please tell me what that test should do. Because this PR adds a check to make sure we don't trip over non-constexpr values in add_ir_attribute_function, there is no way to make it fail.

The test does not need to fail. You can add a test with -verify and // expected-no-diagostics. Like I mentioned in my previous comment, an acceptable test would be one similar to the one you pasted in a comment above. I would suggest not using MockProperties.hpp and spelling out the problematic case in the test itself so it captures what you are trying to fix.

elizabethandrews commented on clang/test/CodeGenSYCL/free_function_implicit_sycl_extern.cpp:
```diff
@@ -0,0 +1,63 @@
+
+// expected-no-diagnostics
+
+// This test checks that specifying SYCL_EXTERNAL is not necessary on a free
```
This test still does not show the FE bug you have fixed. Like I mentioned in the previous comments, please provide a test where `SYCLAddIRAttributesFunction` attribute has the values which caused a crash when used with free functions earlier. I assume you can just write a new test in test/SemaSYCL with the problematic case from `add_ir_attributes_function` (which you have now fixed) to illustrate this.

rdeodhar replied:
_Any_ test that includes sycl.hpp fails without the hasDependent expr check I added. This is because sycl.hpp includes sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp, which has in it:
```
#if defined(__SYCL_DEVICE_ONLY__)
  [[__sycl_detail__::add_ir_attributes_function(
      "sycl-joint-matrix-type", "sycl-joint-matrix-use",
      "sycl-joint-matrix-rows", "sycl-joint-matrix-cols",
      sycl::detail::convertTypeToMatrixTypeString<T>(),
      sycl::detail::convertMatrixUseEnumToString(Use), Rows, Cols)]]
#endif // defined(__SYCL_DEVICE_ONLY__)
```
That amounts to >334 tests.

A test-case with simply `#include <sycl/sycl.hpp>` fails without the check I added. So what is the point of adding yet another test that duplicates the conditions in hundreds of existing tests?

AaronBallman replied:
> So what is the point of adding yet another test that duplicates the conditions in hundreds of existing tests?

I'm confused.

The title of the PR is about removing a need to mark free functions with SYCL_EXTERNAL. The code changes in the PR are talking about constexpr but then implement something about dependency. So we asked you for test coverage to help us understand what it was you were fixing, and now you're pointing us to a constructor in a header file and saying that provides the test coverage. How? It's not a free function.

Also, "this gigantic pile of header files exercises the code path and so that should be sufficient" is really unhelpful to reviewers. We want to look at a test case and go "oh, that's why those changes happened." We should not have to dig through tens of thousands of lines of entirely unrelated header file content for something that can be tested in a self-contained manner. Tests in Clang should have as few includes in them as possible (for a whole host of reasons, including readability).

rdeodhar replied:
The check for a free function now occurs on _every_ function in a compilation unit. The joint matrix header introduces add_ir_attribute_function values that cannot he handled by simply calling `getAttributeNameValuePairs`. Calling hasDependentExpr before filters out the problematic cases.

Do you want to see some atributes added to a free function similar to what is done by the joint matrix header?

Would that modification to the test free_function_implicit_sycl_extern.cpp satisfy the need?

AaronBallman replied:
Oh wait, are there *two* things happening in this review? What I think I'm understanding is:

* We needed to check for dependent arguments to the `SYCLAddIRAttributesFunctionAttr` attribute when processing a declaration; if we don't check for dependent arguments, we'd hit assertions.
* Once we fix that bug, we can treat free functions the same as member functions so that they don't require a special attribute to say it's available on the device.

Is that accurate?

elizabethandrews replied:
>>Is that accurate?
This is my understanding.  The code added in `isFreeFunction` was problematic before because it wasn't doing the dependency check. 

>>Would that modification to the test free_function_implicit_sycl_extern.cpp satisfy the need?
I would suggest removing all headers from that test and writing a simple test which applies the attribute with the previously problematic arguments, on any function which will hit `isFreeFunction`. FE lit testing is not the same as E2E tests. We need to see reduced case with the dependency bug which you have now fixed.

AaronBallman replied:
> This is my understanding.

Thank you!

> I would suggest removing all headers from that test and writing a simple test which applies the attribute with the previously problematic arguments, on any function which will hit isFreeFunction.

+1, based on my understanding, you should be able to write a free function template with a dependent attribute argument and is marked `SYCL_EXTERNAL`; that should reproduce the assertion before your changes, and helps us see that your changes are what removed the assertion.

rdeodhar replied:
In trying to create a test-case that shows the problem with non-constexpr values in add_ir_attribute_function I discovered that the "bad" values only occur when the attribute is applied to a **member function**. See code below.
```
template <typename T> constexpr int value() { return 5; }
template <typename T> struct S {
#if defined(__SYCL_DEVICE_ONLY__)
  [[__sycl_detail__::add_ir_attributes_function("name", value<T>())]]
#endif
  S() {
  }
};
template <typename T>
__attribute__((sycl_device)) [[__sycl_detail__::add_ir_attributes_function(
    "name",
    value<T>())]] [[__sycl_detail__::
                        add_ir_attributes_function("sycl-single-task-kernel",
                                                   0)]] void
f(T i) {}
template void f(int i);
```
For the member function `S()` the attribute looks like this:
```
`-SYCLAddIRAttributesFunctionAttr 0x55989d69e7c0 <col:5, col:67>
  |-ConstantExpr 0x55989d69e870 <col:49> 'const char[5]' lvalue
  | |-value: LValue <todo>
  | `-StringLiteral 0x55989d69e590 <col:49> 'const char[5]' lvalue "name"
  `-CallExpr 0x55989d69e650 <col:57, col:66> '<dependent type>'
    `-UnresolvedLookupExpr 0x55989d69e5d8 <col:57, col:64> '<dependent type>' lvalue (ADL) = 'value' 0x55989d675b30
      `-TemplateArgument type 'T':'type-parameter-0-0'
        `-TemplateTypeParmType 0x55989d675ce0 'T' dependent depth 0 index 0
          `-TemplateTypeParm 0x55989d675c90 'T'
```
while for the function 'f' the attribute looks like this:
```
|-SYCLAddIRAttributesFunctionAttr 0x56361c3c3ea8 <line:37:32, line:39:15>
| |-ConstantExpr 0x56361c3c3f00 <line:38:5> 'const char[5]' lvalue
| | |-value: LValue <todo>
| | `-StringLiteral 0x56361c398cf0 <col:5> 'const char[5]' lvalue "name"
| `-ConstantExpr 0x56361c3c3f60 <line:39:5, col:14> 'int'
|   |-value: Int 5
|   `-CallExpr 0x56361c3c3e88 <col:5, col:14> 'int'
|     `-ImplicitCastExpr 0x56361c3c3e70 <col:5, col:12> 'int (*)()' <FunctionToPointerDecay>
|       `-DeclRefExpr 0x56361c3c3dc0 <col:5, col:12> 'int ()' lvalue Function 0x56361c3c3cc8 'value' 'int ()' (FunctionTemplate 0x56361c398a90 'value')
```
This leads me to believe that the bad values in the attribute will only occur on member functions. Since free functions cannot be member functions filtering out member functions before checking for a function being a free function avoids the problem and is a good idea anyway. With just that change in SemaDecl.cpp all tests pass.
On the other hand, perhaps the bad attribute values could be introduced in some other way. In which case the hasDependentExpr check that I've added in SemaSYCL.cpp would still be necessary.

I've updated the PR to keep both the filtering out of member functions and the hasDependentExpr check. Its a "belt and suspenders" solution. I'd like your opinion whether to keep it that way or remove the hasDependentExpr check.

elizabethandrews commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -16318,11 +16318,15 @@ Decl *Sema::ActOnFinishFunctionBody(Decl *dcl, Stmt *Body,
+  if (FD) {
+    // Free functions cannot be member functions so skip those.
+    const auto *MD = dyn_cast<CXXMethodDecl>(FD);
+    if (LangOpts.SYCLIsDevice && !MD && Body &&
```
Based on your tests below and explanation in PR comments, it sounds like this issue does not have anything to do with `constexpr` values. The issue here is we have non-instantiated functions/methods being checked when they shouldn't be.  Can you try replacing this check with `LangOpts.SYCLIsDevice && Body && FD->isDependentContext()` without the `TemplatedKind` checks. You also do not need the `if (FD)` since L16305 above checks this already.

I think you should also replace the `hasDependentExpr` checks you added in SemaSYCL to asserts instead because we should not be having dependent arguments at that point.

AaronBallman replied:
Slight correction, that should be `!FD->isDependentContext()` -- basically, if the function isn't dependent, it's safe to check the arguments; if the function IS dependent somehow, we need to wait to do those checks until the function has been instantiated (regardless of whether it's a member function or free function).

rdeodhar replied:
The `hasDependentExpr `check I added was a check before calling `getAttributeNameValuePairs` because that would assert if all expressions in the attribute were not constexprs. So there is already an assert there. Perhaps an assert before calling `getAttributeNameValuePairs` is not needed?

Making the other change in SemaDecl.cpp was fine. The tests all pass.

elizabethandrews replied:
Ok. You can remove all that the code you added then if one already exists in  `getAttributeNameValuePairs `

rdeodhar replied:
Done.

elizabethandrews commented on clang/test/Sema/free_function_attribute_check.cpp:
```diff
@@ -0,0 +1,59 @@
+
+// expected-no-diagnostics
+
+// This test checks that non-constexpr values appearing in
```
Can you please update this comment?

You can say something along the lines of "This test ensures that the compiler does not crash when functions or methods with add_ir_attributes_function attribute in dependent contexts are checked for the presence of the free function property"

rdeodhar replied:
OK, I've updated the comment.

rdeodhar commented:
> @rdeodhar tag in PR title should be in `[` and `]`
> 
> https://github.com/intel/llvm/blob/sycl/sycl/doc/developer/ContributeToDPCPP.md#commit-message

OK, done.
